src/Features/Core/Portable/RemoveUnnecessaryParentheses/AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.RemoveUnnecessaryParentheses
+{
+    internal abstract class AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer<
+        TLanguageKindEnum,
+        TParenthesizedExpressionSyntax>
+        : AbstractParenthesesDiagnosticAnalyzer
+        where TLanguageKindEnum : struct
+        where TParenthesizedExpressionSyntax : SyntaxNode
+    {
+        protected AbstractRemoveUnnecessaryParenthesesDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.RemoveUnnecessaryParenthesesDiagnosticId,
+                   new LocalizableResourceString(nameof(FeaturesResources.Remove_unnecessary_parentheses), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                   new LocalizableResourceString(nameof(FeaturesResources.Parentheses_can_be_removed), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        public sealed override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public sealed override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected sealed override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeSyntax, GetSyntaxNodeKind());
+
+        protected abstract TLanguageKindEnum GetSyntaxNodeKind();
+        protected abstract bool CanRemoveParentheses(
+            TParenthesizedExpressionSyntax parenthesizedExpression, SemanticModel semanticModel,
+            out PrecedenceKind precedenceKind, out bool clarifiesPrecedence);
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+        {
+            var syntaxTree = context.SemanticModel.SyntaxTree;
+            var cancellationTokan = context.CancellationToken;
+            var optionSet = context.Options.GetDocumentOptionSetAsync(syntaxTree, cancellationTokan).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var parenthesizedExpression = (TParenthesizedExpressionSyntax)context.Node;
+
+            if (!CanRemoveParentheses(parenthesizedExpression, context.SemanticModel,
+                    out var precedenceKind, out bool clarifiesPrecedence))
```

Comment:
weird indenting

Reply:
it's how we indent to indicate that this is a child of CanRemove.  otherwise, the parameter align with the method itself, and it looks like we have two independent tests.  i.e.:

```c#
if (!Foo() &&
    Bar()
```

vs


```c#
if (!Foo(
        Bar()) &&
```